src/operations/rotate.rs

```diff
@@ -0,0 +1,129 @@
+//! Image rotation operations.
+
+use image::{DynamicImage, ImageBuffer, Pixel, Primitive};
+use imageproc::{definitions::Clamp, geometric_transformations::Interpolation};
+
+use crate::{arg_parsers::RotateGeometry, error::MagickError, image::Image};
+
+fn white_pixel<P>() -> P
+where
+    P: Pixel + Send + Sync,
+    P::Subpixel: Send + Sync + Primitive,
+{
+    let max_value = P::Subpixel::DEFAULT_MAX_VALUE;
+    let channels = P::CHANNEL_COUNT as usize;
+
+    // Create a vector with all channels set to maximum value
+    let subpixels = vec![max_value; channels];
+    P::from_slice(&subpixels).clone()
+}
+
+fn rotate_inner<P>(
+    image: &mut ImageBuffer<P, Vec<P::Subpixel>>,
+    theta: f32,
+) -> Result<(), MagickError>
+where
+    P: 'static + Pixel + Send + Sync,
+    P::Subpixel: Send + Sync + Primitive + Into<f32> + Clamp<f32>,
+{
+    let width = image.width();
+    let height = image.height();
+
+    let (sin_theta, cos_theta) = theta.sin_cos();
+
+    let dst_width = (width as f32 * cos_theta + height as f32 * sin_theta).ceil() as u32;
+    let dst_height = (width as f32 * sin_theta + height as f32 * cos_theta).ceil() as u32;
+
+    //let dst_size = pic_scale_safe::ImageSize::new(
+    //     as usize,
+    //    as usize,
+    //);
+    println!("destination size: {dst_width}x{dst_height}");
+
+    let canvas_width = dst_width.max(width);
+    let canvas_height = dst_height.max(height);
+
+    // Create a larger canvas.
+    let mut canvas = ImageBuffer::from_pixel(canvas_width, canvas_height, white_pixel::<P>());
+
+    image::imageops::overlay(
+        &mut canvas,
+        image,
+        (canvas_width as i64 - width as i64) / 2,
+        (canvas_height as i64 - height as i64) / 2,
+    );
+    *image = imageproc::geometric_transformations::rotate_about_center(
+        &canvas,
+        theta,
+        Interpolation::Nearest,
+        // FIXME: Should account for `-background`.
+        white_pixel::<P>(),
+    );
+
+    let dst_width = dst_width.max(1);
+    let dst_height = dst_height.max(1);
+
+    *image = image::imageops::crop_imm(
+        image,
+        (canvas_width - dst_width) / 2,
+        (canvas_height - dst_height) / 2,
+        dst_width,
+        dst_height,
+    )
+    .to_image();
```

Comment:
Could you add a comment on why we need to expand and then crop the canvas? It's a little difficult to follow for someone who's not familiar with the implementation of this operation.

Does it replicate some imagemagick quirk? Or is it something that's actually always needed for this operation and that `imageproc` doesn't do without all this extra effort from the API user?

Reply:
Sure. The rotation logic in `imageproc` doesn't expand the canvas on its own, as far as I could see. It ends up discarding whatever pixels end up outside the input canvas after the rotation. I will add a comment. 🙂 Or, perhaps, we'd like to contribute to imageproc an appropriate mode of rotation that rotates with expansion?